File: sdks/python/hooksniff/resources/notifications.py
"""
HookSniff SDK — Notifications Resource
"""

from __future__ import annotations

from typing import Any

from ..request import HookSniffRequest, RequestConfig


class Notifications:
    """Notifications & push devices."""

    def __init__(self, config: RequestConfig):
        self._config = config

    def list(self, limit: int | None = None, unread_only: bool = False) -> dict[str, Any]:
        """List notifications."""
        req = HookSniffRequest("GET", "/v1/notifications")
        req.set_query_params({
            "limit": limit,
            "unread_only": unread_only,
        })
        return req.send(self._config, lambda j: j)

    def mark_read(self, notification_id: str) -> dict[str, Any]:
        """Mark a notification as read."""
        req = HookSniffRequest("POST", "/v1/notifications/{id}/read")
        req.set_path_param("id", notification_id)
        return req.send(self._config, lambda j: j)

    def mark_all_read(self) -> None:
        """Mark all notifications as read."""
        req = HookSniffRequest("POST", "/v1/notifications/read-all")
        req.send_no_body(self._config)

    def register_device(self, token: str, platform: str, device_name: str | None = None) -> dict[str, Any]:
        """Register a push notification device."""
        req = HookSniffRequest("POST", "/v1/notifications/devices")
        body: dict[str, Any] = {"token": token, "platform": platform}
        if device_name:
            body["device_name"] = device_name
        req.set_body(body)
        return req.send(self._config, lambda j: j)

    def remove_device(self, device_id: str) -> None:
        """Remove a push notification device."""
        req = HookSniffRequest("DELETE", "/v1/notifications/devices/{id}")
        req.set_path_param("id", device_id)
        req.send_no_body(self._config)

    def get_preferences(self) -> dict[str, Any]:
        """Get notification preferences."""
        req = HookSniffRequest("GET", "/v1/notifications/preferences")
        return req.send(self._config, lambda j: j)

    def update_preferences(self, **kwargs: Any) -> dict[str, Any]:
        """Update notification preferences."""
        req = HookSniffRequest("PATCH", "/v1/notifications/preferences")
        req.set_body({k: v for k, v in kwargs.items() if v is not None})
        return req.send(self._config, lambda j: j)

File: sdks/node/src/resources/notifications.ts
/**
 * HookSniff SDK — Notifications Resource
 */

import { HttpMethod, HookSniffRequest, type HookSniffRequestContext } from "../request";
import type {
  Notification,
  NotificationListResponse,
  NotificationPreferences,
  UpdateNotificationPreferences,
  RegisterDeviceRequest,
  DeviceTokenResponse,
} from "../models";

export class Notifications {
  constructor(private readonly requestCtx: HookSniffRequestContext) {}

  /** List notifications. */
  public list(options?: { limit?: number; is_read?: boolean }): Promise<NotificationListResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/notifications");
    request.setQueryParams({
      limit: options?.limit,
      is_read: options?.is_read,
    });
    return request.send(this.requestCtx, (json) => json as NotificationListResponse);
  }

  /** Get unread count. */
  public unreadCount(): Promise<{ count: number }> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/notifications/unread-count");
    return request.send(this.requestCtx, (json) => json as { count: number });
  }

  /** Mark a notification as read. */
  public markRead(notificationId: string): Promise<Notification> {
    const request = new HookSniffRequest(HttpMethod.PUT, "/v1/notifications/{id}/read");
    request.setPathParam("id", notificationId);
    return request.send(this.requestCtx, (json) => json as Notification);
  }

  /** Mark all notifications as read. */
  public markAllRead(): Promise<void> {
    const request = new HookSniffRequest(HttpMethod.PUT, "/v1/notifications/read-all");
    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Get notification preferences. */
  public getPreferences(): Promise<NotificationPreferences> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/notifications/preferences");
    return request.send(this.requestCtx, (json) => json as NotificationPreferences);
  }

  /** Update notification preferences. */
  public updatePreferences(body: UpdateNotificationPreferences): Promise<NotificationPreferences> {
    const request = new HookSniffRequest(HttpMethod.PUT, "/v1/notifications/preferences");
    request.setBody(body);
    return request.send(this.requestCtx, (json) => json as NotificationPreferences);
  }

  /** Register a device for push notifications. */
  public registerDevice(body: RegisterDeviceRequest): Promise<DeviceTokenResponse> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/devices");
    request.setBody(body);
    return request.send(this.requestCtx, (json) => json as DeviceTokenResponse);
  }
}
